### PR TITLE
extract: use viv-utils.CFG instead of bb.succs

### PR DESCRIPTION
**Pull request by williballenthin**
use `CFG` to compute and access basic block successors. if nothing else, this is about 40% faster than BB succs (which we can remove from viv-utils, again).

also, add documentation explaining the patterns we're looking for.

tested with [b0b758e680e652144a78a7ddecc027d4868c1dc3d8d7d611ec4d3798358b0ce5](https://www.virustotal.com/gui/file/b0b758e680e652144a78a7ddecc027d4868c1dc3d8d7d611ec4d3798358b0ce5) and it seems to extract reasonable strings. 

image

### CODE REVIEW

File: floss/features/extract.py [hunk 1]
```diff
@@ -116,32 +117,75 @@ def extract_function_kinda_tight_loop(f):
...
+            for suc in succs:
+                suc_succs = [x for x in cfg.get_successor_basic_blocks(suc)]
+                if len(suc_succs) != 1:
+                    continue
+                if suc_succs[0] != bb.va:
+                    continue
+
+                loop_bb = suc_succs[0]
+                break
+
```
Comment (williballenthin) [inline floss/features/extract.py:184]: i split out the two cases here to make it more explicit how we check for them

File: floss/features/extract.py [hunk 2]
```diff
@@ -116,32 +117,75 @@ def extract_function_kinda_tight_loop(f):
...
     Algorithm by Blaine S.
     """
+    cfg = viv_utils.CFG(f)
+
+    root_bb_va = cfg.get_root_basic_block()
+    leaf_bb_vas = {bb.va for bb in cfg.get_leaf_basic_blocks()}
+
     for bb in f.basic_blocks:
         # skip first and last BBs
-        if bb.va == f.va or bb.va + bb.size == f.basic_blocks[-1].va:
```
Comment (williballenthin) [inline floss/features/extract.py:121]: bug here that the "last" basic block may not be end of the function. 

now we use the "leaves" that should be the ends of the function.

File: floss/features/extract.py [hunk 3]
```diff
@@ -116,32 +117,75 @@ def extract_function_kinda_tight_loop(f):
     Yields tight loop features in the provided function
     Algorithm by Blaine S.
     """
+    cfg = viv_utils.CFG(f)
```
Comment (mr-tz) [inline floss/features/extract.py:120]: I'm still wondering how I missed this 😄

**Review by mr-tz — APPROVED** [review]
awesome, thanks for the improvement!

**Pull request merged**